apis/k8ssandra/v1alpha1/k8ssandracluster_webhook.go

```diff
@@ -144,6 +147,56 @@ func (r *K8ssandraCluster) ValidateUpdate(old runtime.Object) error {
 	return nil
 }
 
+func validateNoRackRenamed(oldCassandra, newCassandra *CassandraClusterTemplate) error {
+	oldDcs := collectRackNamesPerDc(oldCassandra)
+	newDcs := collectRackNamesPerDc(newCassandra)
+
+	for dc, oldRacks := range oldDcs {
+		if newRacks, ok := newDcs[dc]; ok {
+			if len(newRacks) < len(oldRacks) {
+				return fmt.Errorf("number of racks can't be lowered (DC %s, current=%d, desired=%d)",
+					dc, len(oldRacks), len(newRacks))
+			}
+			for i, oldRack := range oldRacks {
+				newRack := newRacks[i]
+				if newRack != oldRack {
+					return fmt.Errorf("racks can't be renamed (DC %s, index %d, current=%s, desired=%s)",
+						dc, i, oldRack, newRack)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func collectRackNamesPerDc(cassandra *CassandraClusterTemplate) map[string][]string {
+	if len(cassandra.Datacenters) == 0 {
+		// Unlikely but we allow it
+		return map[string][]string{}
+	}
+
+	// The top-level object can define default racks, to be inherited by DCs that don't declare them.
+	// If they aren't declared anywhere, there is a single rack named "default".
+	defaultNames := collectRackNames(cassandra.DatacenterOptions, []string{"default"})
```

Comment:
This logic made me think a bit... We do have a [coalesce](https://github.com/k8ssandra/k8ssandra-operator/blob/main/pkg/cassandra/datacenter.go#L272) function which historically has been used to merge the cluster and DC level configs into a `DatacenterConfig`.

It seems like you're doing some similar logic here, and I wonder if this logic shouldn't be operating on the DatacenterConfig type and doing all validations against a coalesced data structure. At present it seems to be trying to coalesce the cluster and DC level settings, and it might be nicer to re-use the existing logic which does that.